### PR TITLE
Improve the login, chsh and bashrc scripts.

### PR DESCRIPTION
**Pull request by easyaspi314**
#### packages/termux-tools/login:
- The login script can now be called from an empty environment such as from a `su` shell, ADB, or TWRP's terminal.
  - It uses `/system/bin/sh` to run default, because if the user compiles and installs `dash` manually with `--enable-libedit`, we will get the stupid `CANNOT LINK EXECUTABLE` errors in an empty environment. Meanwhile, `/system/bin/sh` seems to work, regardless of `$LD_LIBRARY_PATH`.
  - It makes sure `/system/bin/linker` exists and aborts if it doesn't, prompting the user to mount `/system` (the likely reason for this).
  - It makes sure that `$PREFIX`, `$HOME`, `$LD_LIBRARY_PATH`, and `$PATH` are set unless the user sets `$TERMUX_PRESERVE_ENVIRONMENT`.
  - It notifies the user if they are running as root and provides a hint about how to fix permissions, as mucking around as root tends to create files owned by root and it sometimes breaks SELinux contexts.
  - It includes `/sbin/sh`, the TWRP shell, if necessary, to use as a login shell. This should never happen as /system should be mounted, but whatever.

#### packages/bash/etc-bash.bashrc:
- "Now more useful than /system/etc/mkshrc!™"
- There is now a nice shell prompt instead of a boring `$ `. It looks like this, in color if possible:

img_20180602_211342

  - It will change `/data/data/com.termux/files` (well, technically, `$PREFIX/..`) to `termux_root` to make things shorter.
  - If the user is root, instead of saying `termux`, it will say `root` in red.
  - If the user is not on localhost, it will say `user@hostname` in magenta.
- `ls` is in pretty colors now.
- It makes a few popular aliases.
  - All `busybox` utilities can now be called by prefixing `bb` to their name for short, such as `bbls`.
  - It aliases `nano` to `nano -w` unless the user manually configured `nano`. Because, let's face it, `nano`'s wrapping never helped anyone.
- It sets some helpful `bash` options.
- It sets `$EDITOR` to `vi` by default.

### CODE REVIEW

**Comment by easyaspi314** [comment]
Ok, can someone tell me what is wrong with packages/termux-tools/install.sh? I have no idea why it isn't working. :cry:

**Comment by Grimler91** [comment]
TERMUX_HOME isn't set, use TERMUX_ANDROID_HOME

**Comment by Grimler91** [comment]
Regarding 

> It uses /system/bin/sh to run default, because if the user compiles and installs dash manually with --enable-libedit, we will get the stupid CANNOT LINK EXECUTABLE errors in an empty environment.

If the user compiled dash and get that error, then the user didn't compile dash correctly. I don't think you need to work around this hypothetical situation, so I would probably keep $PREFIX/bin/sh as shebang.

The [fail-safe mode](https://wiki.termux.com/wiki/Recover_a_broken_environment) uses /system/bin/sh so users with a faulty shell can still use fail-safe mode to fix their environment.

**Comment by easyaspi314** [comment]
On the contrary. Dash has an optional `--with-libedit` feature that links with libedit.0.so.

After patching mktokens to replace /tmp with $TMPDIR, do this
```
$ apt install libedit libedit-dev ldd
$ ./configure --prefix=$PREFIX --with-libedit
$ make
$ ldd src/dash
libedit.so.0
libdl.so
libc.so
```
Meanwhile, unless I am mistaken, /system/bin/sh will only need libc, libc++, and libdl.

**Comment by Grimler91** [comment]
Ah, so using /system/bin/sh is to make sure all shells without LD_LIBRARY_PATH can use the script. I didn't understand that at first.
Sounds more legit.

File: packages/termux-tools/login [hunk 1]
```diff
@@ -1,13 +1,82 @@
-#!/data/data/com.termux/files/usr/bin/sh
+#!/system/bin/sh
+# ^ If the user compiles dash by hand with libedit, this will fail
+# if LD_LIBRARY_PATH is set.
```
Comment (Grimler91) [inline packages/termux-tools/login]: Typo here? "If LD_LIBRARY_PATH is not set"?

Comment (easyaspi314) [reply packages/termux-tools/login]: Good catch.

I'm like a spell check. I can tell when a word is spelled wrong or whether I used "whom" properly, but I have trouble noticing when something is just plain wrong. :(

I'll fix it when I get home.

**Comment by easyaspi314** [comment]
I made a few improvements to login.
* I fixed the typo.
* It will now actually check for /system/bin/linker as well as for /system/lib[64]/libc.so instead of being commented out.
  * It will now give a more helpful message to TWRP users if this happens, as /system is unmounted by default and TWRP doesn't make this entirely clear.
* Termux will use ~/.termux/shell if it is executable, not part of the same GID. That had to be a typo from ones before me. This lets you use, say, /system/bin/sh as a login shell and presumably, something other than bash on a root login.
* The scripts makes absolutely sure $SHELL is set by checking /proc/$$/exe to try and get the shell currently interpreting the login script as a last resort, and failing if it still can't find it.
* I use echo now instead of printf. For some strange and stupid reason, mksh, and by extension, /system/bin/sh, lack a printf built-in.
* A few other minor changes.

**Comment by easyaspi314** [comment]
@fornwall do you have any thoughts?

**Comment by easyaspi314** [comment]
I also made a slight tweak to chsh that allows the user to choose any shell by an absolute path.

**Comment by easyaspi314** [comment]
@fornwall, apologies about the double ping, but it has been about two weeks since any activity and I just want to remind you that I really want to hear your feedback. 

If you want me to make any changes, I will gladly try to do so. It might take a while (I am currently using an unholy iPhone until I get a battery situation sorted out), but I can either try to find my old S II Skyrocket or do stuff not specifically related to Termux on my MacBook. 

Anyways, if you are not interested in this, please let me know as well. I will understand and close this. 

Don’t feel like I need an immediate reply; I have plenty of time before I can get out of iOS hell. I understand if you’re busy. Thanks.

**Comment by tomty89** [comment]
No offense, but IMHO this PR isn't really appropriate. Most of the stuff (if not all) are really personalization. Personally (again, no offense) I don't want this to be merged.

I think all we need is to mark `bash.bashrc` and `login` as conf file.

**Comment by easyaspi314** [comment]
Almost all distros have a bashrc (or whatever rcfile name is for the default shell) that contain settings and personalizations they think are useful. By their nature, they can be overridden in `~/.bashrc` and/or `~/.bash_profile` if the user dislikes them, or would like to tweak the setup to their needs.  

For example, here are a few:
* [Android](https://android.googlesource.com/platform/external/mksh/+/jb-dev/mkshrc)
* [LineageOS (for their copy of bash they provide)](https://github.com/LineageOS/android_external_bash/blob/lineage-15.1/etc/bashrc)
* [Debian](https://sources.debian.org/src/bash/5.0%7Ealpha1-1/debian/etc.bash.bashrc/)
* [macOS](https://opensource.apple.com/source/bash/bash-106/bashrc.auto.html). Okay, maybe not the best example. 
* [Cygwin](https://cygwin.com/git/gitweb.cgi?p=cygwin-apps/base-files.git;a=blob;f=etc/defaults/etc/bash.bashrc;h=0f633fa725ccf2dee9cc5e74d58052a2a891c9d7;hb=HEAD)
* [Arch Linux](https://git.archlinux.org/svntogit/packages.git/tree/trunk/system.bashrc?h=packages/bash)
* [OpenBSD](https://github.com/openbsd/src/blob/master/etc/ksh.kshrc)

I don’t think it is fair that we give the user, who intentionally chose to use a terminal environment on their phone, a next-to-useless shell profile when Android itself provides a more useful profile on a shell that was only really meant to be used rarely via USB. 

If we want to mimic a Unix environment, we should at least provide a useful shell profile.

**Comment by tomty89** [comment]
I sincerely don't see how the defaults of bash is "next to useless". And why would you expect everyone to like those one-character aliases (and even a loop to make tons of aliases for busybox). (Since you mentioned Arch, its bash ships only *one* alias, which is not necessary in Termux AFAIK.) Not to mention the shopts that quietly changes the default/expected behavior of the shell.

You added "optimization" for `nano` and preferred `vi(m)`, what about emacs and others? ;-) 

Why would I want an extra .dot directory and file being written on every shell start when I am not even using a rooted ROM?

> By their nature, they can be overridden in ~/.bashrc and/or ~/.bash_profile if the user dislikes them, or would like to tweak the setup to their needs.

Why shouldn't we be allowed to (not) set up the environment as we desired?

**Comment by Quasic** [comment]
Maybe this could be a package or something, but I don't know about putting it in the main repo. I like a bare shell and try to rip out such things. That's one reason I like this app, but it's also very convenient and helpful to know what is useful customization for this system, so I think this could be used somewhere else.

**Pull request closed without merging**